eclipse-theia/theia
# watchers: use singleton server

Theia spawns as many nsfw servers as there are clients, maybe more.

This commit replaces this behaviour by centralizing everything into the
same service process in order to optimize resources (watch handles are a
limited resource on a given Linux host).

Some notes on the current design:
- Every time someone requests to watch an `(uri, options)` couple we will
re-use the same watch handle (and actual nsfw instance). This allows for
resource allocation optimizations.
- Since a given service can only serve a single client, I added an event
dispatcher named `FileSystemWatcherServerDispatcher`. You must register
yourself with some unique id you will use to make requests.
- The new `.watchUriChanges` function now takes a `clientId` so that
when events are sent to the dispatcher it then gets routed back to your
own client.
- Added a `--nsfw-watcher-verbose` backend argument to enable verbose
logging for the watcher server.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>



#### How to test

- Enable the `sample.file-watching.verbose` preference.
- Run `yarn --cwd examples/browser start --nsfw-watcher-verbose`
- Watch for the different logs coming out of the watcher server:
  - There shouldn't be new watchers without a frontend opened.
  - Once you open a frontend on a given workspace for the first time, you should see `NEW` log notifications.
  - If you open a second frontend on the same workspace, you should see `REF++` log notifications
  - Closing one of the two frontend should emit `REF--` log notifications.
  - Closing the last frontend should schedule the watcher de-allocation about 10s later.
    - But if you re-open a frontend before 10s, it should reference the watchers again and not dispose of them.
- Comparing with master:
  - Run the backend.
  - Run `lsof | grep inotify | wc -l` a couple of times before opening a frontend and note the general value.
  - Open frontends on some big workspace (your home maybe?) while noting the new value for `lsof | grep inotify | wc -l` each time.
  - The conclusion should be that on master the number increases with each frontend, but not with these changes.
- Running the `@theia/example-browser` app:
  - Open 2 tabs on different workspaces.
  - Create and delete files in the respective workspaces.
  - Make sure that a log is only reported from the frontend that was watching the modified workspace.
    - `Sample File Watching: X file(s) changed! <workspace>`

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
